### DISCUSSION TITLE
Sometime I received an Exception Response: IllegalFunction | IllegalAddress | GatewayNoResponse

### DISCUSSION THREAD

**Discussion by zarch** [pymodbus-dev/pymodbus#2389] [Q&A] [answered]
Hi all,

I'm using pymodbus to read values from a bunch of devices via modbus tcp.

Looking at the log, I see that sometime the device raise some exceptions.
Bellow I reported the exceptions logs from about 10 minutes of test,  please consider that I'm sending a request to the registers every 30 seconds.

**SLAVE 1**

| time                    | error             | register | count | slave |
|-------------------------|-------------------|----------|-------|-------|
| 2024-10-16 08:22:02.156 | GatewayNoResponse | 20522    | 20    | 1     |
| 2024-10-16 08:22:31.098 | IllegalFunction   | 20522    | 20    | 1     |
| 2024-10-16 08:23:01.081 | IllegalFunction   | 20522    | 20    | 1     |
| 2024-10-16 08:23:31.100 | IllegalFunction   | 20522    | 20    | 1     |
| 2024-10-16 08:25:01.085 | IllegalFunction   | 20522    | 20    | 1     |
| 2024-10-16 08:25:31.100 | IllegalFunction   | 20522    | 20    | 1     |
| 2024-10-16 08:27:02.150 | GatewayNoResponse | 20522    | 20    | 1     |

**SLAVE 2**
| time                    | error             | register | count | slave |
|-------------------------|-------------------|----------|-------|-------|
| 2024-10-16 08:22:32.098 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:23:02.081 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:23:32.099 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:24:32.095 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:25:02.085 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:25:32.100 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:26:02.084 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:28:02.087 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:29:02.096 | IllegalFunction   | 20480    | 35    | 2     |
| 2024-10-16 08:31:32.088 | IllegalFunction   | 20480    | 35    | 2     |
|                             |                       |               |         |       |
| 2024-10-16 08:22:04.155 | GatewayNoResponse | 20522    | 20    | 2     |
| 2024-10-16 08:22:33.098 | IllegalAddress    | 20522    | 20    | 2     |
| 2024-10-16 08:23:03.081 | IllegalAddress    | 20522    | 20    | 2     |
| 2024-10-16 08:23:33.099 | IllegalAddress    | 20522    | 20    | 2     |
| 2024-10-16 08:24:03.101 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:24:33.096 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:25:03.087 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:25:33.101 | IllegalAddress    | 20522    | 20    | 2     |
| 2024-10-16 08:26:03.085 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:27:03.076 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:28:03.087 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:29:03.096 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:31:03.095 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:31:33.088 | IllegalFunction   | 20522    | 20    | 2     |
| 2024-10-16 08:32:03.093 | IllegalFunction   | 20522    | 20    | 2     |

and **SLAVE 3**

| time                    | error             | register | count | slave |
|-------------------------|-------------------|----------|-------|-------|
| 2024-10-16 08:22:05.301 | GatewayNoResponse | 20480    | 35    | 3     |
| 2024-10-16 08:23:04.080 | IllegalFunction   | 20480    | 35    | 3     |
| 2024-10-16 08:24:04.101 | IllegalFunction   | 20480    | 35    | 3     |
| 2024-10-16 08:29:04.096 | IllegalAddress    | 20480    | 35    | 3     |
| 2024-10-16 08:30:05.164 | GatewayNoResponse | 20480    | 35    | 3     |
| 2024-10-16 08:30:35.174 | GatewayNoResponse | 20480    | 35    | 3     |
| 2024-10-16 08:31:04.095 | IllegalFunction   | 20480    | 35    | 3     |
| 2024-10-16 08:31:34.087 | IllegalFunction   | 20480    | 35    | 3     |
| 2024-10-16 08:32:04.094 | IllegalFunction   | 20480    | 35    | 3     |
|                             |                       |               |         |       |
| 2024-10-16 08:23:05.081 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:24:05.101 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:24:35.096 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:27:05.076 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:27:35.084 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:30:05.238 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:30:35.249 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:31:05.094 | IllegalAddress    | 20522    | 20    | 3     |
| 2024-10-16 08:32:05.094 | IllegalFunction   | 20522    | 20    | 3     |
| 2024-10-16 08:32:36.167 | GatewayNoResponse | 20522    | 20    | 3     |

As you can see the requests are not raised every 30s and indeed I'm acquiring some data! :-)
I can understand the `GatewayNoResponse` issue that might be due to some connectivity problem (even if everything is cabled). What it is hard to unsderstand to me is why the same requests can raise sometime an `IllegalFunction` or an `IllegalAddress`.

What can explain this nondeterministic device's behavior? How can I further investigate this issue?

Here a single request raising an `IllegalFunction` with pymodbus log level to `DEBUG`:

```log
2024-10-16 08:12:01,028 DEBUG transaction:66 Adding transaction 14
2024-10-16 08:12:01,028 DEBUG transport:379 send: 0x0 0xe 0x0 0x0 0x0 0x6 0x1 0x4 0x50 0x2a 0x0 0x14
2024-10-16 08:12:01 [T:MainThread] DEBUG:pymodbus.logging: Adding transaction 14
2024-10-16 08:12:01 [T:MainThread] DEBUG:pymodbus.logging: send: 0x0 0xe 0x0 0x0 0x0 0x6 0x1 0x4 0x50 0x2a 0x0 0x14
2024-10-16 08:12:01,101 DEBUG transport:322 recv: 0x0 0xe 0x0 0x0 0x0 0x3 0x1 0x84 0x1 old_data:  addr=None
2024-10-16 08:12:01,102 DEBUG base:101 Processing: 0x0 0xe 0x0 0x0 0x0 0x3 0x1 0x84 0x1
2024-10-16 08:12:01,102 DEBUG factory:259 Factory Response[132]
2024-10-16 08:12:01,102 DEBUG base:119 Frame advanced, resetting header!!
2024-10-16 08:12:01,103 DEBUG transaction:77 Getting transaction 14
2024-10-16 08:12:01.103 | CRITICAL | Exception Response(132, 4, IllegalFunction) | client=pymodbus.client.tcp.AsyncModbusTcpClient | func_type='input_registers' | address=20522, count=20, slave=1
```

**Comment by janiversen** [comment, marked as answer]
As you can see from the debug log, it is your device that reports IllegalFunctiomRequest, so pymodbus is ok.

Looking at the previous send, it is a legal modbus message, but maybe your device do not like the address or the amount of data you requiere.
